Add northbound API proto spec

kuujo opened the pull request:
This PR is an initial attempt at adding a Protobuf spec for the northbound API. A few notes on the design decisions and then you can feel free to debate them...

First, I feel the use of explicit `*Request` and `*Response` messages should become a best practice for Protobuf specs. The reason is they guard against future changes. We can only modify the messages - e.g. add a filter to `GetDevicesRequest` - if the request is a `message`. If the request were instead a `string`, we'd be limited forever to it being a `string`. So, that's the reason for explicit request/response messages, and it's a pattern I've followed strictly in [Atomix Protobuf APIs](https://github.com/atomix/atomix-api).

Second, this spec provides separate RPCs for getting a list of devices and for streaming device changes. I think there's probably a debate to be had about the behavior of `StreamDevices`. There's a benefit in terms of consistency to the RPC streaming all existing devices plus future devices changes. In other words, if the store has 3 devices when the request is received, stream `ADD` response for each of the three devices plus all future device events. The benefit is this avoids either missing device events or duplicate device events, and it ensures only a single request is needed. If the `StreamDevices` RPC simply returns future device events, then the client must first `GetDevices` and then `StreamDevices`, and this risks missing or duplicate events between the two streams. For this reason, `StreamDevicesRequest` supports a flag to stream the store in `ADD` events plus later events.

Finally, because the onos-topo service is a remote service, we have to change the treatment of secrets. Secrets must be accessible on whichever node is accessing the device, so the `TlsConfig` simply provides the filename for the CA cert, cert, and key. These will have to be distributed by k8s and available on a secret path on each node which needs to access the device. In production, the user will provide a secret to the Helm chart, and the Helm chart will mount the secret to each of the ONOS nodes so that when a device is read and the secret file name is provided, each node can find the secret file.

There's also a question of naming the messages that are not `*Request` and `*Response` messages - e.g. `Device`, `Credentials`, and `TlsConfig`. One could argue that they should be prefixed or suffixed with an identifier indicating they're distinct from the structs we'll probably want to create in our own Go code, but I'd argue we can preserve the simple names as the Golang package naming strategy namespaces them anyways, i.e. code will reference `pb.Device` as an object distinct from e.g. `admin.Device`.

ccascone commented:
Is the idea to support the management of not just devices but other entities as well? If so I think we could raise the level of abstraction by using generic Write / Read / Stream RPCs, instead of entity-specific ones, like it is done for P4Runtime.

For example, instead of having an `AddDevice` RPC with `AddDeviceRequest` argument, we could have a `Write` RPC with `WriteRequest` argument defining type `ADD` (or `UPDATE` or `DELETE`) and value a `Device` message (or a `Link` message, or any other topology entity we need to manage). Similarly, we could have a generic Stream RPC to deliver any kind of event, not just device-related ones, reducing the amount of code duplication at the server and client.

Here's the P4Runtime WriteRequest message:
https://github.com/p4lang/p4runtime/blob/master/proto/p4/v1/p4runtime.proto#L59

Toughts?

kuujo commented:
I’m not really opposed to that approach. I’d leave it up to the experts who better understand the breadth of use cases and contexts this API needs to support. @tomikazi should think about this

SeanCondon commented on pkg/northbound/proto/admin.proto:
```diff
@@ -19,6 +19,152 @@ syntax = "proto3";
+    }
+
+    // GetDevices gets a list of devices
+    rpc GetDevices(GetDevicesRequest) returns (GetDevicesResponse) {
```
the reply here could be stream too (rather than a synchronous response)

ccascone replied:
Agreed, the response should be a stream, as it makes reading a large number of devices much more efficient. Probably the use cases we are trying to cover do not include a large number of devices, but it seems a nice design decision that would reduce tech debt. Especially if we go down the path of using generic Read RPC (cc @tomikazi). Say one day we will need to get Links, or Ports, or Channels (of a port)... that can easily become a large number. I am strongly in favor of using a generic Read RPC (with type `DEVICE`, 'LINK', etc) as we automatically enforce the same good design for every read operation. instead of worrying whether we need a stream response or not.

tomikazi replied:
Yes, I completely agree that the get response should be a stream. I also agree that there will be additional entities down the line. I was envisioning the topology subsystem as a collection of services for tracking devices, their ports, links, networks.

I'm not opposed to collapsing this, but reductio ad absurdum will lead us to an API with mere GET/PUT. To me this is a trade-off in readability and maintanability. The repeated patterns in the methods in the various inventory services (device, link, networks, etc.) are a down-side, but on the other side where we have a common facade for accessing everything will be an intertwined and complex set of messages and schemes to differentiate one type of entity from another. In my view, while it will admittedly result in a slightly repeated pattern between devices, link, networks, etc., it will also give independence to the different services to properly cater to their respective subjects, in addition to making things more readable.

kuujo replied:
It's not really clear to me what the benefit of streaming the list of devices is. Unless we anticipate the device list as being so large that it has the potential to impact memory/efficiency, streaming doesn't make sense for a request-reply message that's returning a discrete set of objects. This is _why_ the `repeated` field exists. Streaming the response increases the complexity of clients which then have to aggregate a response that really should have already been aggregated by the gRPC framework. Streams should be reserved for data that is a legitimate _stream_, and when they're not the specification is just describing a discrete set as a continuous one - describing the API and increasing the complexity of clients - without much benefit that I can see.

And I tend to agree with @tomikazi on the point that reducing the RPCs using `oneof` creates a more complicated message structure. What is the benefit beyond reducing the total number of RPCs, which seem to be basically free? I would suggest also that the API should be a collection of services as opposed to a single service for all entities.

ccascone replied:
The benefit is scalability. Using streams now allow us to consider the case where we need to return and process a large number of devices (and links, ports, channels, etc) from the get-go. Moreover, I think the additional complexity at the client for the aggregation part is negligible, and shouldn't be a concern if this design favors scalability. What is more, grpc-java offers natively the possibility of wrapping a streaming RPC like this one in a sync stub, handling the aggregation of responses internally. I believe other grpc implementations for other languages provide the same capability. If that's not the case, apps can always implement their internal libraries to perform aggregation and re-use that.

Using generic Read/Write RPC with `oneof` field(s) is a way to follow the [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) principle. I sketched here what I have in mind: https://gist.github.com/ccascone/32fc537f70c56ab927b67f5c69444d74

Does it make sense to apply it if all we need to manage is 3 different entities? Maybe not, what about 10? or 20? Perhaps knowing the type of entities that need to be managed would help us make an informed decision.

tomikazi replied:
These are all good points. I need to think about it a bit more. It could be my intuition is off. :)

Andrea-Campanella replied:
My two cents. 

1. I definitely like the streaming, independently of the approach we choose for the API methods/entities, so that should be adopted. 
2. Regarding the entities and method calls discussion. I do like https://gist.github.com/ccascone/32fc537f70c56ab927b67f5c69444d74 from @ccascone.
We can go for this solution if we think a first class object is an optical link vs a ran link for example. 
In this case the number of entities I think it will be in the order of 30-50 considering IP, Optical, RAN traffic. 
if instead we prefer to have a generic `link` first class object and then have `link_type` attribute the number of object will probably reduce to 10~, at least that's my impression. 
My preference would be to be as explicit as possible, thus creating first class objects for different types of devices, links etc. thus picking the solution Carmelo proposed.

SeanCondon commented on pkg/northbound/proto/admin.proto:
```diff
@@ -19,6 +19,152 @@ syntax = "proto3";
+
+// GetDevicesRequest gets a list of devices in the topology
+message GetDevicesRequest {
+
```
Could have a name filter maybe?

SeanCondon commented on pkg/northbound/proto/admin.proto:
```diff
@@ -19,6 +19,152 @@ syntax = "proto3";
+// received by the service. However, if `all` is `true`, the stream is prefixed by a device `ADD` event
+// for each existing device in the store.
+message StreamDevicesRequest {
+
```
Could have a name filter maybe

tomikazi replied:
Are you suggesting within? If so, yes... among other criteria such as type, availability, etc.

kuujo replied:
I think that’s fine. We should address filtering and device parameters separately. The `Device` doesn’t currently even include most of the parameters on which we’d want to filter, but it is consistent with what’s in onos-config currently. I think we should either keep it that way until we’ve integrated onos-config and onos-topo and then address the changes needed to `Device` and filtering, or if someone wants to propose a change to the `Device` fields we can merge this and then add that in another PR before the onos-config/onos-topo work.

Andrea-Campanella commented:
I like Carmelo's approach because the topology API could explode easily with the complexity of objects due to the combination of RAN, L2/L3, L0 etc.

tomikazi commented:
I see where P4RT is coming from, but in this case, I think we're already talking about a fairly bounded set of methods and for app readability my preference would be to stick with the Get/Add/Delete style.

I think the outer messages should all be *Request, *Response and if there are filter-related entities within, those should be named that. But the outer message should be *Request.

tomikazi commented:
> I like Carmelo's approach because the topology API could explode easily with the complexity of objects due to the combination of RAN, L2/L3, L0 etc.

The device would remain generic and would be open-ended type/enum. So adding new types of entities should not explode the API.

kuujo commented:
I've updated the spec. A couple of changes:
1. A collection of `Device`s can be retrieved via the `GetAll` RPC. 
  * All devices present in the store when the stream is opened will be streamed with the `PRESENT` type, indicating the `Device` was already present in the store when the request was received. For a request where `subscribe` is `false`, this will be the `type` of all response messages. 
  * When `subscribe` is `true`, the stream will remain open after all `PRESENT` devices have been streamed to the client, and events that occur after all existing devices have been streamed will be sent to the client with either `ADDED`, `UPDATED`, or `REMOVED` events, allowing the client to build a state machine/cache from events.
2. I also added an `Update` RPC. The rationale for this is that `Add` and `Update` RPCs need to be treated very differently by the server, especially to preserve consistency. They can just as easily be combined in a single write RPC where the device is either added or updated depending on whether the `device` field is empty. However, if we do so then we need a coherent name for the upsert RPC: `AddOrUpdate`?
3. Regardless of whether we merge `Add`/`Update`, the API also needed some metadata from the stores themselves which the clients can use for concurrency control. I added the `ObjectMetadata` message for this purpose. `ObjectMetadata` is a generic message which can be reused for all the entity types and contains an ID and version which can be used by the server to prevent concurrent changes or changes to outdated objects.

Andrea-Campanella commented:
> 2. I also added an `Update` RPC. The rationale for this is that `Add` and `Update` RPCs need to be treated very differently by the server, especially to preserve consistency. They can just as easily be combined in a single write RPC where the device is either added or updated depending on whether the `device` field is empty. However, if we do so then we need a coherent name for the upsert RPC: `AddOrUpdate`?

I think we can merge the two, like in the ONOS classic device store we had AddOrUpdate method we could do the same here
